main.c

```diff
@@ -11,22 +12,33 @@ int main() {
   const char proc_path[] = "/proc";
   struct dirent **name_list;
   int i, dir_count;
-  const int max_proc = get_max_proc();
+  int err;
+  int max_proc;
+
+  err = get_max_proc(&max_proc);
+  if (err) {
+    printf("Error while obtaining max_proc\n");
+    return 1;
+  }
+
 
   dir_count = scandir(proc_path, &name_list, is_proc_dir, numsort);
   if (dir_count < 0) {
         perror("scandir failed"); // TODO: more detailed error description
+        return 1;
   }
-
   ProcessNode_t *root_node = create_process_node(NULL);
-  ProcessNode_t *current_node;
-  Process_t *process;
+  ProcessNode_t *current_node = NULL;
+  Process_t *process = NULL;
   ProcessNode_t **node_array = malloc(sizeof(ProcessNode_t) * max_proc);
-  ProcessNode_t *parent_node;
+  ProcessNode_t *parent_node = NULL;
   node_array[0] = root_node;
 
   for(i = 0; i < dir_count; i++) {
```

Comment:
Think twice about the scope of i variable. If the variable is defined outside of the loop scope it will safe value. Do you need the "i" outside loop?